### ISSUE TITLE
Random test failures `test_model_navigation_indexer_slice`

### ISSUE THREAD

**Issue by vidartf** [hyperspy/hyperspy#914]
I sometimes encounter the following failure in tests. The matter only occurs sometimes, and rerunning the test will often succeed. Possibly a race condition? Do we have anything that relies on the order in which events/traits notification is processed?

```python
======================================================================

FAIL: hyperspy.tests.model.test_fancy_indexing.TestModelIndexing.test_model_navigation_indexer_slice

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda/envs/testenv/lib/python3.5/site-packages/nose/case.py", line 198, in runTest

    self.test(*self.arg)

  File "/home/travis/build/hyperspy/hyperspy/hyperspy/tests/model/test_fancy_indexing.py", line 81, in test_model_navigation_indexer_slice

    assert_false(m[0]._active_array[0, 0])

AssertionError: True is not false
```

**Comment by francisco-dlp** [comment]
travis is failing in #919 because of this.

**Comment by vidartf** [comment]
Rerunning travis in #919 caused it to succeed.

**Comment by sem-geologist** [comment]
What exactly this test is testing?
the ```m[0]._active_array[0, 0]``` looks for me as slice of some kind of array in array? (forgive me I am not familiar with hyperspy models). However, looking at the TestModelIndexing.setUp method I see that there is created numpy array with float type (dtype is not specified, so by default it is float). If I understand correct then the ```m[0]._active_array[0, 0]``` will return random float number. So it does not surprise me that it randomly fails.

Generally speaking the test looks suspicious, there are more lines which can fail in the future, in example ```assert_true(np.all(m.channel_switches[1:] == True))```.
```m.channel_switches[1:]``` probably returns tuple or list. Those should not be compared with True.

**Comment by vidartf** [comment]
From memory, active array should be array of booleans. All the times the test fails, it says `True is not false`, which to me indicates that the `dtype` is correct (since `True` is boolean), bu that the value is wrong (should be `False`).

**Comment by sem-geologist** [comment]
Ok, I went through the test  in python command line checking it step by step. the final array results with [0,0] element beeing False on my computer, so it does pass the test. generally It looks to over complicated for me. I tried to track down what is going on and got to _signals/spectrum_simulation.py :open_mouth:  Why somebody would get so abstract....?

**Comment by to266** [comment]
FYI the `spectrum_simulation.py` here is most likely irrelevant - it's only used to generate the noise.

I agree that `np.testing.*` should have been used, I just was not aware of the functions at the time of writing. If I remember correctly, as I went through a number of implementations of the slicing, these test, if working correctly, should be either comparing arrays that actually point to the same memory (hence identical by definition), or created using `np.copy` or something along these lines. It should be safe for now. It's trivial to replace the tests with numpy-supplied ones if it starts to fail.

Your worries for the `channel_switches` slicing are also misplaced - it's a numpy array of booleans, slicing it returns a numpy array, and here we rely on numpy to do the broadcasting. Its an extremely common piece of code, I'm surprised you have not encountered it before..
Actually there should not even be a comparison, just `m.channel_switches[1:].all()`

As for the race conditions, obviously the relevant lines are these
```python
# ...
self.model.axes_manager.indices = (0, 0)
self.model[0].active = False
m = self.model.inav[0::2]
```
Here I explicitly change the current index to `(0,0)` and rely on the [`_set_active`](https://github.com/hyperspy/hyperspy/blob/master/hyperspy/component.py#L732) to record the change to the `_active_array`, and it _does_ involve both traits and events... Though I can't say I can spot the race condition right now.

**Comment by to266** [comment]
https://github.com/hyperspy/hyperspy/pull/959 Should help with this. Forgot to tag with PR...

**Comment by vidartf** [comment]
@to266: I reran the Travis test in #959 a couple of times, and got it to fail, so while it might have been improved, it is at least not fixed completely yet. That shouldn't prevent #959 from going through, it just won't close this issue.

**Comment by to266** [comment]
I was afraid it did not work... thanks any way

**Comment by vidartf** [comment]
Should be fixed in #1067, reopen if not.
